pkg/volume/rbd/rbd.go

```diff
@@ -357,6 +362,142 @@ func (plugin *rbdPlugin) ConstructVolumeSpec(volumeName, mountPath string) (*vol
 	return volume.NewSpecFromVolume(rbdVolume), nil
 }
 
+// <><><>
+
+func (plugin *rbdPlugin) ConstructBlockVolumeSpec(podUID types.UID, volumeName, mapPath string) (*volume.Spec, error) {
+	pluginDir := plugin.host.GetVolumeDevicePluginDir(rbdPluginName)
+	blkutil := volutil.NewBlockVolumePathHandler()
+
+	client := plugin.host.GetKubeClient()
+	globalMapPathUUID, err := blkutil.FindGlobalMapPathUUIDFromPod(pluginDir, mapPath, podUID)
+	if err != nil {
+		return nil, err
+	}
+	glog.V(5).Infof("globalMapPathUUID: %v, err: %v", globalMapPathUUID, err)
+	globalMapPath := filepath.Dir(globalMapPathUUID)
+	if len(globalMapPath) == 1 {
+		return nil, fmt.Errorf("failed to retreive volume plugin information from globalMapPathUUID: %v", globalMapPathUUID)
+	}
+	return getVolumeSpecFromGlobalMapPath(volumeName /* globalMapPath , */, client)
+}
+
+func getVolumeSpecFromGlobalMapPath(volumeName string /* globalMapPath string , */, client clientset.Interface) (*volume.Spec, error) {
+	// Retreive volume spec information from globalMapPath
+	// globalMapPath example:
+	//   plugins/kubernetes.io/{PluginName}/{DefaultKubeletVolumeDevicesDirName}/{volumePluginDependentPath}
+	pvToRecover, err := client.CoreV1().PersistentVolumes().Get(volumeName, metav1.GetOptions{})
```

Comment:
@sbezverk can you check how often this get called?

Reply:
The volume reconstruction only happens if following two conditions are satisfied.
- Kubelet node is down
- Pod(and volume) is deleted during node down

When kubelet node is coming back, volume reconstruction happens.
Therefore this method doesn't happen many times, but if PV is deleted together with Pod, this logic doesn't work. I guess we need fall back logic if API can't be reached or PV is deleted.

Reply:
If API is not reachable I think non of the command getting any k8s objects would work and most likely this would cause greater issues than non-reconstructed PV. When I have a normal cluster I will try this scenario.

Reply:
- If API server is down or kubelet cannot reach API server, failure to reconstruct a rbd PV is not bad - In case the Pod is deleted in the API server during disconnection without kubelet awareness, failure to get PV from API server avoids mounting a rbd image that might be re-assigned or owned by another Pod.
- If PV is deleted, failure to get PV just shifts the timing of hitting the problem: Since PV is deleted, reconstruct won't be able to bring up the mount. Falling back to use global path will then hit a mount problem during pod recreation.

Reply:
@rootfs 

What I saw during volume reconstruction and following sequences are,
(1) reconstruct a volume using construct method of volume plugin
(2) Container is terminated
(3) dswp removes the volume from dsw
(4) unmount and unmountDevice happen for the volume

Does mounting a rbd image happen during these sequence?

>> If API server is down or kubelet cannot reach API server, failure to reconstruct a rbd PV is not bad

I agree this is OK since volume reconstruction happens periodically(per 3-6 min).